### PR TITLE
Changing the credentials plugin API.

### PR DESCRIPTION
**Pull request by jboeuf**
The plugin is now passed more information that it can use to create
  auth metadata:
    - service_url (as before)
    - method name
    - channel_auth_context

Also adding a type field that can be used later if we want to set policies for which credentials are incompatible during composition.

### CODE REVIEW

**Comment by jboeuf** [comment]
@jtattermusch @murgatroid99 @yang-g @tbetbetbe @stanley-cheung @ctiller @soltanmm @nathanielmanistaatgoogle please have a look. Comments are welcome.

This is the API change only so that you can see what's coming. I'm starting to work on the implementation and will adjust if you want me to change the API.

**Comment by jboeuf** [comment]
An example of how the new API looks in C++ is here: 5eaeb82.

File: include/grpc/grpc_security.h [hunk 1]
```diff
@@ -165,6 +240,23 @@ typedef void (*grpc_credentials_plugin_metadata_cb)(
     void *user_data, const grpc_metadata *creds_md, size_t num_creds_md,
     grpc_status_code status, const char *error_details);
 
+/* Context that can be used by metadata credentials plugin in order to create
+   auth related metadata. */
+typedef struct {
```
Comment (murgatroid99) [inline include/grpc/grpc_security.h]: Why do we need all of this information in the metadata plugin? Why do we need the auth context?

Comment (jboeuf) [inline include/grpc/grpc_security.h]: You may have some credentials that want to bind some token to the server's channel identity.

**Comment by jboeuf** [comment]
@mvrable may want to have a look as well.

**Comment by jtattermusch** [comment]
The changes look alright to me, but it looks like we might need to adjust the metadata plugin API in wrapped languages so that the argument passed to the auth plugin callback is an object (of class "AuthMetadataContext" or some other fitting name). We won't expose the newly added fields like "method" or "auth_context" because currently we don't have use for them (correct?), but we need to make sure we can expose them with only additive API change if we need make them available in the future. Once this PR is merged, can you pls file issues for all wrapped languages to cover this (or comment on the existing issues for languages where the auth work hasn't been done yet)?

I'd still like a review for the additions to the C core public api (grpc_security.h). @yang-g , can you please take a look?

Btw, I'm OOO next week, so I will reassign to @murgatroid99 .

**Comment by jboeuf** [comment]
Assigning @ctiller for the core API review. Will give it back to @murgatroid99 after we're in agreement.

**Comment by yang-g** [comment]
LGTM

**Comment by jboeuf** [comment]
for @yang-g Re: fully qualified method name. I added some documentation that should make things clearer.

**Comment by jboeuf** [comment]
The errors are due to jenkins's internal flakyness. Merging.

**Pull request merged**